### DISCUSSION TITLE
What is the way to structure state with similar properties?

### DISCUSSION THREAD

**Discussion by bnn1** [pmndrs/zustand#949] [General]
I have a state `headerState` that controls two UI elements: Mobile navbar and Auth dialog.  

Mobile navbar and auth dialog both have properties `isOpen`, `open` and `close`

I need to come up with a flexible and extendible solution on how to structure the state.

I see several ways of solving this problem:

```javascript
type NoArgsNoReturn = () => void;
// structure by category
type HeaderState = {
  navMenu: { isOpen: boolean; open: NoArgsNoReturn; close: NoArgsNoReturn },
  authDialog: { isOpen: boolean; open: NoArgsNoReturn; close: NoArgsNoReturn }
};

// structure by what they do
type HeaderState = {
  isOpen: { navMenu: boolean; authDialog: boolean },
  open: { navMenu: NoArgsNoReturn; authDialog: NoArgsNoReturn },
  close:{ navMenu: NoArgsNoReturn; authDialog: NoArgsNoReturn }
};

// separate states
type HeaderState = {
  isNavMenuOpen: boolean;
  isAuthDialogOpen: boolean;
  // if `open` is provided, isNavMenuOpen will be set to `open`, otherwise it will be toggled
  setNavMenuOpen: (open?: boolean) => void; 
  // same for `isAuthDialogOpen`
  setAuthDialogOpen: (open?: boolean) => void; 
}
```

But I can't figure out what would be the best way to structure it, there must be some best practices or practices developed from experience. 

What do you think is the way to go? Why?

**Comment by dai-shi** [comment]
(looks like a good discussion topic. :eyes:)

**Reply by bnn1** [reply]
Hope so :)) What's your opinion on the matter?

**Reply by dai-shi** [reply]
Since the original pattern for zustand is [Flux inspired practice](https://github.com/pmndrs/zustand/wiki/Flux-inspired-practice), and that's why zustand merges state (only the first level), I'd go with option 3. But, I understand option 1 might play well for some large scale apps. In such a case, we'd need to merge state (the second level) by ourselves, or use some third-party library for namespacing.
